junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/TestTemplateTestDescriptor.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2015-2017 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v1.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.junit.jupiter.engine.descriptor;
+
+import static java.util.stream.Collectors.toList;
+import static org.junit.platform.commons.meta.API.Usage.Internal;
+
+import java.lang.reflect.Method;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.junit.jupiter.api.extension.ContainerExtensionContext;
+import org.junit.jupiter.api.extension.TestTemplateInvocationContext;
+import org.junit.jupiter.api.extension.TestTemplateInvocationContextProvider;
+import org.junit.jupiter.engine.execution.JupiterEngineExecutionContext;
+import org.junit.jupiter.engine.extension.ExtensionRegistry;
+import org.junit.platform.commons.meta.API;
+import org.junit.platform.commons.util.Preconditions;
+import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.engine.TestTag;
+import org.junit.platform.engine.UniqueId;
+import org.junit.platform.engine.support.descriptor.MethodSource;
+import org.opentest4j.TestAbortedException;
+
+/**
+ * {@link TestDescriptor} for {@link org.junit.jupiter.api.TestTemplate @TestTemplate}
+ * methods.
+ *
+ * @since 5.0
+ */
+@API(Internal)
+public class TestTemplateTestDescriptor extends JupiterTestDescriptor {
+
+	private final Class<?> testClass;
+	private final Method templateMethod;
+
+	public TestTemplateTestDescriptor(UniqueId uniqueId, Class<?> testClass, Method templateMethod) {
+		super(uniqueId, determineDisplayName(Preconditions.notNull(templateMethod, "Method must not be null"),
+			MethodTestDescriptor::generateDefaultDisplayName));
+
```

Comment:
Feature envy and duplication have been addressed in 628f14b03b5382055eace6e79e790e24396e80cf.